plugins/catalog/src/components/SystemDiagram/SystemDiagram.tsx

```diff
@@ -0,0 +1,123 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Entity } from '@backstage/catalog-model';
+import { catalogApiRef } from '@backstage/plugin-catalog-react';
+import {
+  DependencyGraph,
+  DependencyGraphTypes,
+  InfoCard,
+  Progress,
+  useApi,
+} from '@backstage/core';
+import { Alert } from '@material-ui/lab';
+import React from 'react';
+import { useAsync } from 'react-use';
+
+type SystemDiagramProps = {
+  entity: Entity;
+  variant?: string;
+};
+
+export function SystemDiagram({ entity }: SystemDiagramProps) {
+  const catalogApi = useApi(catalogApiRef);
+  const { loading, error, value: catalogResponse } = useAsync(() => {
+    return catalogApi.getEntities({
+      filter: {
+        kind: ['Component', 'API', 'Resource', 'System', 'Domain'],
+      },
+    });
+  }, [catalogApi]);
+
+  const currentSystemName = entity.metadata.name;
+  const currentSystemNode = `system:${entity.metadata.name}`.toLowerCase();
+  const systemNodes = [];
+  const systemEdges = [];
+
+  if (catalogResponse && catalogResponse.items) {
+    for (const catalogItem of catalogResponse.items) {
+      // pick out the system itself
+      if (catalogItem.metadata.name === entity.metadata.name) {
+        systemNodes.push({
+          id: currentSystemNode,
+        });
+
+        // check if the system has an assigned domain
+        // even if the domain object doesn't exist in the catalog, display it in the map
+        if (catalogItem.spec?.domain) {
+          systemNodes.push({
+            id: `domain:${catalogItem.spec.domain}`.toLowerCase(),
+          });
+          systemEdges.push({
+            from: currentSystemNode,
+            to: `domain:${catalogItem.spec.domain}`.toLowerCase(),
+            label: 'part of',
+          });
+        }
+      }
+
+      // process any entity assigned to the system
+      if (catalogItem.spec?.system === currentSystemName) {
+        systemNodes.push({
+          id: `${catalogItem.kind}:${catalogItem.metadata.name}`.toLowerCase(),
+        });
+
+        // check relations of the entity to see if it relates to other entities
+        // note those relations may, or may not, be explicitly
+        // assigned to the system
+        if (catalogItem.relations) {
+          for (const relation of catalogItem.relations) {
+            switch (relation.type) {
+              case 'providesApi':
+                systemEdges.push({
+                  to: `${catalogItem.kind}:${catalogItem.metadata.name}`.toLowerCase(),
+                  from: `${relation.target.kind}:${relation.target.name}`.toLowerCase(),
```

Comment:
You do this a lot - maybe make an `getEntityNodeId` utility function or so?

Also you skip namespace - hm should you not either filter by namespace, or include it in the id? Could envision deps to things in other namespaces, whose nodes could be like colored differently or something :)

Reply:
So again, apols TypeScript novice here.  I threw in this anon function:

```tsx
const getEntityNodeId = ({ entity }: { entity: Entity }) => {
  return (
    `${entity.kind}:${entity.metadata.name}`.toLowerCase()
  );
}
```

But when I call it, I'm getting a type error:

```tsx
systemNodes.push({
  //id: `${catalogItem.kind}:${catalogItem.metadata.name}`.toLowerCase(),
  id: getEntityNodeId(catalogItem),
});
```

> Argument of type 'Entity' is not assignable to parameter of type '{ entity: Entity; }'.ts(2345)

Suggestions?   I've checked TypeScript docs, Google, StackOverflow, and it's not obvious to me yet.

Reply:
Yep your function expects an argument that is an object (`{}`) with one field named `entity`, but you give an Entity instead.

Declare your function as follows instead:

```ts
const getEntityNodeId = (entity: Entity) => {
```
 
Is actually using function :$

```ts
function getEntityNodeId(entity: Entity) {
```

*grumbles* you kids and your newfangled arrow functions and your hippity-hop

Reply:
Works great for the main entity notes, but haven't gotten it working for `relations.target` yet, though also likely relates to your point about `getEntityRelations()`. So will hack at that a bit later.